services/filter.service.js

```diff
@@ -128,59 +206,153 @@ function FilterService(options) {
+          jsonobj[realKey] = getConvertValueByTypeKeyname(obj[key][key1], val, listfields,param);
+          if (jsonobj[realKey] === undefined) {
+            jsonRes = undefined;
+            return false;
+          }
+          if (nbkey < 2) {
+            jsonval = jsonobj;
+          }
+          else {
+            var jsonobj2 = {};
+            jsonobj2 = jsonobj;
+            jsonval["$or"].push(jsonobj);
+          }
+          return true;
+        });
+        if (jsonRes === undefined) {
+          return undefined;
+        }
+        jsonRes = jsonval;
       }
       else {
-        jsonRes[getRealKeyNeeded(key, obj[key])] = getConvertValueByTypeKeyname(obj[key], val, listfields, errors);
+        jsonRes[getRealKeyNeeded(key, obj[key], obj, param)] = getConvertValueByTypeKeyname(obj[key], val, listfields, param);
+        jsonRes = addObjOptionIfNeeded(jsonRes, key, obj, param );
       }
     }
     return jsonRes;
   }
   
-  function getConditionAfter(val, obj, listfields, errors) {
+  function getConditionAfter(val, obj, listfields, param) {
     var jsonRes = {};
     var value;
-    Object.keys(obj).forEach(function (key) {
-      if (key[0] === '$') {
-        jsonRes = fillJsonObjWithKeyAfter(jsonRes, key, val, obj, listfields, errors);
+    Object.keys(obj).every(function (key) {
+      if (key in SPECIAL_OPERATORS) {
+        jsonRes[key] = getSpeOPeratorVal(key, obj[key], param);
+        if (jsonRes[key] === undefined) {
+          jsonRes = undefined;
+          return false;
+        }
+      }
+      else if (key[0] === '$') {
+        jsonRes = fillJsonObjWithKeyAfter(jsonRes, key, val, obj, listfields, param);
+        if (jsonRes === undefined) {
+          return false;
+        }
       }
       else if ((value = getValIfValExistInArray(key, listfields)) !== undefined)
       {
-        jsonRes[value] = (getConditionAfter(value, obj[key], listfields, errors));
+          jsonRes[value] = (getConditionAfter(value, obj[key], listfields, param));
+        if (jsonRes[value] === undefined) {
+          jsonRes = undefined;
+          return false;
+        }
       }
       else {
-        errors.push("key: "+ key + " need to be a condition started with '$'");
+        param["errors"].push("key: "+ key + " need to be a condition started with '$'");
         jsonRes = undefined;
-        return undefined;
+        return false;
       }
+      return true;
     });
     return (jsonRes);
   }
   
-  function getConditionBeforeArray(currentKey, array, listfields, errors) {
+  function getConditionBeforeArray(currentKey, array, listfields, param) {
     var jsontab = [];
     for(var i = 0; i < array.length; i++) {
       var value = array[i];
-      if (typeof value !== 'object') {
+      if (typeof value !== 'object' && (currentKey === undefined || currentKey === null))
+      {
+        param["errors"].push("cond : " + JSON.stringify(value) + " can't be before the field name");
```

Comment:
@bosulliv42 
Same thing, please enclose dynamic values with double-quotes.
And replace "cond" with "conditional structure" (it is more explicit).

Info: In English, there is no spaces between the character ":" and the preceding word :  
`hello: world` and not `hello : world`